text/0048-schema-annotations.md

```diff
@@ -0,0 +1,131 @@
+# Annotations for Cedar Schemas 
+
+## Related issues and PRs
+
+- Reference Issues: 
+- Implementation PR(s): https://github.com/cedar-policy/cedar/tree/feature/shaobo/rfc48
+
+## Timeline
+
+- Started: 2024-02-05
+
+## Summary
+
+Like Cedar policies, users may want to associate arbitrary, machine readable metadata with Schema objects.
+We solved this problem in Cedar policies by allowing for *annotations*: arbitrary key/value pairs that are attachable to policies.
+This could be extended to Cedar schemas, allowing users to attach attributes an entity type, common type, and action definitions.
+
+
+## Basic example
+
+Here is a basic example for doc comments.
+```
+@doc("This entity defines our central user type")
+entity User { 
+    manager : User,
+    team : String
+};
+```
+The `@id("...")` notation is similar to the notation used for policy annotations.
+
+## Motivation
+
+Users should be allowed to associate machine readable metadata with objects in a Schema.
+While we could create special syntax for associating particular kinds of metadata, we cannot
+predict all of the metadata uses that users will have. 
+Thus providing a flexible system that users can adapt to their needs is preferable.
+This proposal re-uses the same syntax from Cedar Policies, creating a unified syntax.
+
+
+## Detailed design
+
+### Semantics
+Attributes have **no** impact on validation decisions. 
+Attributes are arbitrary key/value pairs where:
+* 'key' is a valid Cedar identifier
+* 'value' is a Cedar string
+
+
+The Cedar spec takes no opinion or stance on the interpretation of annotations.
+The interpretation is entirely up to users of Cedar.
+
+### Cedar Schema Syntax
+Attributes in Cedar Schemas will mirror the syntax used for attributes in a policy: informally that's `@<key>("value")`.
+Formally the following rule is added to the Cedar grammar: 
+```
+Annotation := '@' IDENT '(' STR ')'
+Annotations := {Annotations}
+```
+With an arbitrary number of them being able to prepend to a top level declaration in a schema.
+
+Thus the full schema syntax becomes:
+```
+Schema      := {Namespace}
+Namespace   := ('namespace' Path '{' {Decl} '}') | {Decl}
```

Comment:
Based on some recent discussion, we might want to support annotations on the namespace. I don't see a reason not to.